iroh-util/src/lib.rs

```diff
@@ -60,8 +61,9 @@ pub async fn block_until_sigint() {
 /// | macOS    | `$HOME`/Library/Application Support/iroh   | /Users/Alice/Library/Application Support/iroh |
 /// | Windows  | `{FOLDERID_RoamingAppData}`/iroh           | C:\Users\Alice\AppData\Roaming\iroh   |
 pub fn iroh_config_root() -> Result<PathBuf> {
-    let cfg = dirs_next::config_dir()
-        .ok_or_else(|| anyhow!("operating environment provides no directory for configuration"))?;
+    let cfg = dirs_next::config_dir().ok_or(UtilError::NoDir {
+        purpose: "configuration data",
+    })?;
```

Comment:
I'm not really sure I like this style of using `UtilError` much.  This is now an error type that's a grab bag of errors that can happen for functions in this module.  It does help you handling the error from e.g. this function itself since you have to guess (I mean, the typesystem doesn't help you) which error variants you should handle for this function.  So it's not really any better then using `anyhow::Error`.

As for error ergonomics, I'd expect the error to preserve the underlying error.  It's all good saying there's no directory, but was that because the directory entry did not exist or because there was no read permission on the direntry?

To me a pretty reasonable way of improving this error message is to do something like `dirs_next::config_dir().context("no directory for configuration")?;`.  For something that can raise a single error variant only there's fairly low value in starting to provide custom error types, it would only wrap the single underlying error.  What you want to do is enrich it with the usage information, which is what anyhow's `Context` was made for.

Reply:
> This is now an error type that's a grab bag of errors that can happen for functions in this module. 

I disagree that it's a grab bag. It's an enumeration of the types of errors this module can emit. If it looks like a grab bag, it's probably because the module API is poorly constructed.

It does help you handling the error from e.g. this function itself since you have to guess (I mean, the typesystem doesn't help you) which error variants you should handle for this function.


>  As for error ergonomics, I'd expect the error to preserve the underlying error.

Yes totally. I'll fix this. It's wrong to swallow the underlying error. I also thing your point generalizes, we should always be propagating the underlying error.

>  For something that can raise a single error variant only there's fairly low value in starting to provide custom error types, it would only wrap the single underlying error. 

Up in application developer space, sure. But the motivation for this change is to make Iroh easier to consume as a library. Aside from not forcing our consumers to import `anythow`, the benefit of this change making a more explicit contract about the kinds of errors an iroh library will emit.

I like anyhow for application developers, as a library developer I find it lazy, and as a user of libraries that return `anyhow::Result` I've already encountered situations where `anyhow` is just too vague to be useful. I needed to make this conversion with `LockError` prior to our v0.1.0 release to be able to handle problems with program locks (which are common), and [provide feedback on missing services](https://github.com/n0-computer/iroh/blob/dc6f85925f9ad4681d0a647a1803180bedde28ee/iroh/src/main.rs#L41-L54).

But all that said, this is an important decision to make as a team. This overhead of enumerating error types will slow down the dev process, and I think there's a case to be made for doing this slowly over time with APIs as we stabilize them. The counterpoint is similar to our windows support story, not doing the work as we go sets us up for a massive amount of catch up work later, and may leave us exposed to having to undo decisions & revisit problems